src/main/java/com/diffplug/gradle/CmdLine.java

```diff
@@ -143,42 +140,32 @@ public static Result runCmd(File directory, String cmd, boolean echoCmd, boolean
 		Process process = builder.start();
 
 		// wrap the process' input and output
-		try (
-				BufferedReader stdInput = new BufferedReader(new InputStreamReader(process.getInputStream(), Charset.defaultCharset()));
-				BufferedReader stdError = new BufferedReader(new InputStreamReader(process.getErrorStream(), Charset.defaultCharset()));) {
-
+		try {
 			if (echoCmd) {
 				System.out.println("cmd>" + cmd);
 			}
 
-			// dump the output
-			ImmutableList.Builder<String> output = ImmutableList.builder();
-			ImmutableList.Builder<String> error = ImmutableList.builder();
-
-			String line = null;
-			while ((line = stdInput.readLine()) != null) {
-				output.add(line);
-				if (echoOutput) {
-					System.out.println(line);
-				}
-			}
-
-			// dump the input
-			while ((line = stdError.readLine()) != null) {
-				error.add(line);
-				if (echoOutput) {
-					System.err.println(line);
-				}
-			}
+			InputStreamCollector stdInputThread = new InputStreamCollector(process.getInputStream(), echoOutput ? System.out : null, null);
+			stdInputThread.start();
+			InputStreamCollector stdErrorThread = new InputStreamCollector(process.getErrorStream(), echoOutput ? System.err : null, null);
+			stdErrorThread.start();
 
 			// check that the process exited correctly
 			int exitValue = process.waitFor();
-			if (exitValue != EXIT_VALUE_SUCCESS) {
+			// then wait for threads collecting the output of thread
+			stdInputThread.join();
+			stdErrorThread.join();
+
+			if (stdInputThread.getException() != null) {
+				throw stdInputThread.getException();
+			} else if (stdErrorThread.getException() != null) {
+				throw stdErrorThread.getException();
+			} else if (exitValue != EXIT_VALUE_SUCCESS) {
 				throw new RuntimeException("'" + cmd + "' exited with " + exitValue);
 			}
 
```

Comment:
Just noted this issue: Maybe it's not the best idea to re-throw exceptions from another thread without wrapping them, since the stacktrace of the exception won't match to the thread that throws it :-/